File: java/SlidingWindow_java/Minimum_Window_Substring_76.java
package java.SlidingWindow_java;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Minimum_Window_Substring_76 {
    public String minWindow(String s, String t) {
        HashMap<Character, Integer> mp = new HashMap<>();

        int i, j, count;
        String ans = "";
        for (i = 0; i < t.length(); i++) {
            mp.put(t.charAt(i), mp.getOrDefault(t.charAt(i), 0) + 1);
        }
        count = mp.size();
        for (i = 0, j = 0; i < s.length(); i++) {
            mp.put(s.charAt(i), mp.getOrDefault(s.charAt(i), 0) - 1);
            if (mp.get(s.charAt(i)) == 0)
                --count;

            if (count == 0) {
                while (count == 0) {
                    if (ans.length() > i - j + 1 || ans.isEmpty()) {
                        ans = s.substring(j, i + 1);
                    }
                    mp.put(s.charAt(j), mp.get(s.charAt(j)) + 1);
                    if (mp.get(s.charAt(j)) == 1)
                        ++count;
                    ++j;
                }
            }
        }
        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Minimum_Window_Substring_76 obj = new Minimum_Window_Substring_76();
            // int nums[] = { 1,2,3,4,7 };
            op.println(obj.minWindow("ADOBECODEBANC", "ABC"));

        }
    }

}

File: Code/SlidingWindow/Minimum_Window_Substring_76.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    string minWindow(string s, string t)
    {
        unordered_map<char, int> mp;
        int i, j, count;
        string ans = s;
        for (i = 0; i < t.size(); i++)
        {
            mp[t[i]]++;
        }
        count = mp.size();
        for (i = 0, j = 0; i < s.size(); i++)
        {
            mp[s[i]]--;
            if (mp[s[i]] == 0)
                --count;

            if (count == 0)
            {
                while (count == 0)
                {
                    if (ans.size() > i - j + 1)
                    {
                        ans = s.substr(j, i - j + 1);
                    }
                    ++mp[s[j]];
                    if (mp[s[j]] == 1)
                        ++count;
                    ++j;
                }
            }
        }
        return ans;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.minWindow("ADOBECODEBANC", "ABC");
    cout << ans;
    return 0;
}